src/cert.rs

```diff
@@ -101,6 +101,10 @@ pub(crate) fn parse_cert_internal<'a>(
         // special logic for handling critical Netscape Cert Type extensions.
         // That has been intentionally omitted.
 
+        if !tbs.peek(der::Tag::ContextSpecificConstructed3 as u8) {
+            return Err(Error::UnsupportedCertWithoutExtensions);
+        }
+
```

Comment:
I don't think we should spend 4 lines of code on this, and we shouldn't do the extra peek.

I think instead, you can change the call to `der::nested_mut` below to get the same effect, similar to your version parsing patch.

Reply:
I tried, seems like it is not possible to change call to `nested_mut` to handle this case correctly without significant rewrite of both webpki and ring: `der::nested_mut` takes only one error parameter, which handles all the cases:
* tag is different from requested — only here we should return `UnsupportedCertWithoutExtensions`
* tagged data is inconsistent with tag
* content parser did not consume all tagged data

Passing `UnsupportedCertWithoutExtensions` to `der::nested_mut` will result in webpki reporting `UnsupportedCertWithoutExtensions` on supported but incorrect DER.

Reply:
> I tried, seems like it is not possible to change call to `nested_mut` to handle this case correctly without significant rewrite of both webpki and ring: `der::nested_mut` takes only one error parameter, which handles all the cases:

Yes, that's actually a problem with the other PR regarding version numbers too, although I guess that one is affected less.

I think we can extend the *ring* API to include a new variant of `expect_tag_and_get_value` that returns different errors for mismatched tags vs. invalid value.

Reply:
I tried rewriting with if, looks like this: https://github.com/stepancheg/webpki/commit/d75b4ed3540f34733e71bdf0a928cf315525ff0b

The `if` is now inside `expect_tag_and_get_value`.

Patching `ring` to return different errors will be ever harder, because `ring` error has no variants.

What can I do next?

Reply:
Thinking about this overnight, I think we should just compromise and make some incremental progress:

```
            der::nested(
                tbs,
                der::Tag::ContextSpecificConstructed3,
-               Error::BadDER,
+              Error::MissingOrMalformedExtensions,
```

Whether the extensions are missing or malformed really doesn't matter since the result is the same: the certificate is rejected, but the more specific error will help narrow down where to look.